src/cmake/dependency_utils.cmake

```diff
@@ -0,0 +1,377 @@
+#
+# After running, it leaves the following variables set:
+#   ${pkgname}_LOCAL_SOURCE_DIR
+#   ${pkgname}_LOCAL_BUILD_DIR
+#   ${pkgname}_LOCAL_INSTALL_DIR
+#
+# Unless NOINSTALL is specified, the after the installation step, the
+# installation directory will be added to the CMAKE_PREFIX_PATH and also will
+# be stored in the ${pkgname}_ROOT variable.
+#
+macro (build_dependency_with_cmake pkgname)
+    cmake_parse_arguments(_pkg   # prefix
+        # noValueKeywords:
+        "NOINSTALL"
+        # singleValueKeywords:
+        "GIT_REPOSITORY;GIT_TAG;VERSION"
+        # multiValueKeywords:
+        "CMAKE_ARGS"
+        # argsToParse:
+        ${ARGN})
+
+    message (STATUS "Building local ${pkgname} ${_pkg_VERSION} from ${_pkg_GIT_REPOSITORY}")
+
+    set (${pkgname}_LOCAL_SOURCE_DIR "${${PROJECT_NAME}_LOCAL_DEPS_ROOT}/${pkgname}")
+    set (${pkgname}_LOCAL_BUILD_DIR "${${PROJECT_NAME}_LOCAL_DEPS_ROOT}/${pkgname}-build")
+    set (${pkgname}_LOCAL_INSTALL_DIR "${${PROJECT_NAME}_LOCAL_DEPS_ROOT}/dist")
+    message (STATUS "Downloading local ${_pkg_GIT_REPOSITORY}")
+
+    # Clone the repo if we don't already have it
+    find_package (Git REQUIRED)
+    if (NOT IS_DIRECTORY ${${pkgname}_LOCAL_SOURCE_DIR})
+        execute_process(COMMAND ${GIT_EXECUTABLE} clone ${_pkg_GIT_REPOSITORY}
+                                -b ${_pkg_GIT_TAG} --depth 1
+                                ${${pkgname}_LOCAL_SOURCE_DIR})
+        if (NOT IS_DIRECTORY ${${pkgname}_LOCAL_SOURCE_DIR})
+            message (FATAL_ERROR "Could not download ${_pkg_GIT_REPOSITORY}")
+        endif ()
+    endif ()
+    execute_process(COMMAND ${GIT_EXECUTABLE} checkout ${_pkg_GIT_TAG}
+                    WORKING_DIRECTORY ${${pkgname}_LOCAL_SOURCE_DIR})
+
+    set (_pkg_quiet OUTPUT_QUIET)
+
+    # Configure the package
+    execute_process (COMMAND
+        ${CMAKE_COMMAND}
+            # Put things in our special local build areas
+                -S ${${pkgname}_LOCAL_SOURCE_DIR}
+                -B ${${pkgname}_LOCAL_BUILD_DIR}
+                -DCMAKE_INSTALL_PREFIX=${${pkgname}_LOCAL_INSTALL_DIR}
+            # Same build type as us
+                -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+            # Shhhh
+                -DCMAKE_MESSAGE_INDENT="        "
+                -DCMAKE_MESSAGE_LOG_LEVEL=WARNING
+                -DCMAKE_COMPILE_WARNING_AS_ERROR=OFF
+                -DCMAKE_VERBOSE_MAKEFILE=OFF
+                -DCMAKE_RULE_MESSAGES=OFF
+            # Build args passed by caller
+                ${_pkg_CMAKE_ARGS}
+        ${pkg_quiet}
+        )
+
+    # Build the package
+    execute_process (COMMAND ${CMAKE_COMMAND}
+                        --build ${${pkgname}_LOCAL_BUILD_DIR}
+                     ${pkg_quiet}
+                    )
+
+    # Install the project, unless instructed not to do so
+    if (NOT _pkg_NOINSTALL)
+        execute_process (COMMAND ${CMAKE_COMMAND}
+                            --build ${${pkgname}_LOCAL_BUILD_DIR}
```

Comment:
On multiconfig generators (MSVC) `CMAKE_BUILD_TYPE` is [ignored](https://discourse.cmake.org/t/cmake-build-type-presets-multi-config-generators/9699). In this case we should specify the build type with `--config` parameter to avoid mixing release and debug builds.
```Imath-3_1_d.lib(half.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '2' doesn't match value '0' in argparse.obj ```

Reply:
Yes, thanks, I believe I've already done that in the version I'm currently using/editing, but I haven't pushed to have it append to this PR for several days. I still am having some Windows issues I was trying to work through before updating.

Reply:
I hesitated on passing the same `CMAKE_BUILD_TYPE` while building required dependencies - as these are well tested external dependencies and in majority cases building them in Release mode should be preferred. Do you think `--config` should take a separate build type to avoid long reconfiguration times if the main `CMAKE_BUILD_TYPE` gets modified?

Reply:
I think you have a good point. I believe there are two use cases here: (1) even when building a debug version of OIIO, you want optimized dependencies because those are dependable anyway; (2) you're doing a deep debug and the problem may indeed be one where you need to trace into a dependency.

I think you're right, though, that (1) is probably the more common case. So maybe I'll add a separate control for the build mode for any locally build dependencies.

Or maybe it's moot? I propose the following hypothesis: the kind of users who need and will use the auto-built dependencies are probably doing a one-time build of our package that they're maybe not too familiar with, whereas the kinds of people who are likely to want a debug build of OIIO are probably more sophisticated users who already have separately installed copies of the dependencies. The strong circumstantial evidence is that anybody building OIIO over and over again probably doesn't want a fresh download and build to happen for dependencies every time they clean the build area of their OIIO. So it's also possible that this doesn't really matter, because there is rarely an intersection between people doing debug builds and who need the on-the-fly dependency builds. 

I'll add some control for this once I work out my Windows woes.